lib-extra/src/jdt/java/com/diffplug/spotless/extra/glue/jdt/EclipseJdtCoreManipulation.java

```diff
@@ -0,0 +1,258 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.extra.glue.jdt;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.eclipse.core.internal.runtime.InternalPlatform;
+import org.eclipse.core.resources.IFolder;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IProjectDescription;
+import org.eclipse.core.resources.ProjectScope;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.ILog;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.content.IContentTypeManager;
+import org.eclipse.core.runtime.preferences.DefaultScope;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences;
+import org.eclipse.jdt.core.IBuffer;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IPackageFragment;
+import org.eclipse.jdt.core.IPackageFragmentRoot;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.manipulation.CodeStyleConfiguration;
+import org.eclipse.jdt.core.manipulation.JavaManipulation;
+import org.eclipse.jdt.internal.core.BufferManager;
+import org.eclipse.jdt.internal.core.CompilationUnit;
+import org.eclipse.jdt.internal.core.DefaultWorkingCopyOwner;
+import org.eclipse.jdt.internal.core.JavaCorePreferenceInitializer;
+import org.eclipse.jdt.internal.core.PackageFragment;
+import org.eclipse.jdt.internal.core.nd.indexer.Indexer;
+import org.eclipse.jdt.internal.corext.codemanipulation.CodeGenerationSettingsConstants;
+
+import com.diffplug.spotless.extra.eclipse.base.SpotlessEclipseFramework;
+
+/**
+ * Basic set-up for formatting features based on {@code org.eclipse.jdt:org.eclipse.jdt.core.manipulation}.
+ */
+class EclipseJdtCoreManipulation {
+	// The JDT UI shall be used for creating the settings (JavaUI not imported due to dependencies).
+	private final static String JDT_UI_PLUGIN_ID = "org.eclipse.jdt.ui";
+
+	private final static String ROOT_AS_SRC = "";
+	private final static String PROJECT_NAME = "spotless";
+	private final static String SOURCE_NAME = "source.java";
+
+	protected final ILog logger;
+
+	private final AtomicInteger uniqueProjectId = new AtomicInteger(0);
+	private final Map<String, String> defaultOptions;
+
+	protected EclipseJdtCoreManipulation() throws Exception {
+		if (SpotlessEclipseFramework.setup(
```

Comment:
I need to check the migration of EclipseJdtFormatterStepImpl to see how this was migrated.

Reply:
This `SpotlessEclipseFramework` stuff is tricky. Depending on what version of Eclipse, and what methods you are calling, it might not matter if any of the OSGi stuff has been initialized at all. Or you might need all of it initialized.

For the current Eclipse JDT and CDT formatters, nothing is initialized at all, and that works fine. For Groovy we needed to start the full Solstice OSGi thing. If you are getting NPE or "blah not initialized" errors, I would copy this into your static initializer as a template.

https://github.com/diffplug/spotless/blob/bc21f833c000b4095232bf211edacd8450a7ac2d/lib-extra/src/groovy/java/com/diffplug/spotless/extra/glue/groovy/GrEclipseFormatterStepImpl.java#L52-L70

Reply:
Thanks for the insights. For now, I'm feeling a bit overwhelmed here. I'll take some time to get through it.